cfg.c

```diff
@@ -0,0 +1,245 @@
+    cfg->expand = 1;
+  } else if (strncmp(arg, "userpresence=", 13) == 0) {
+    sscanf(arg, "userpresence=%d", &cfg->userpresence);
+  } else if (strncmp(arg, "userverification=", 17) == 0) {
+    sscanf(arg, "userverification=%d", &cfg->userverification);
+  } else if (strncmp(arg, "pinverification=", 16) == 0) {
+    sscanf(arg, "pinverification=%d", &cfg->pinverification);
+  } else if (strncmp(arg, "authfile=", 9) == 0) {
+    cfg->auth_file = arg + 9;
+  } else if (strcmp(arg, "sshformat") == 0) {
+    cfg->sshformat = 1;
+  } else if (strncmp(arg, "authpending_file=", 17) == 0) {
+    cfg->authpending_file = arg + 17;
+  } else if (strncmp(arg, "origin=", 7) == 0) {
+    cfg->origin = arg + 7;
+  } else if (strncmp(arg, "appid=", 6) == 0) {
+    cfg->appid = arg + 6;
+  } else if (strncmp(arg, "prompt=", 7) == 0) {
+    cfg->prompt = arg + 7;
+  } else if (strncmp(arg, "cue_prompt=", 11) == 0) {
+    cfg->cue_prompt = arg + 11;
+  } else if (strncmp(arg, "debug_file=", 11) == 0) {
+    const char *filename = arg + 11;
+    debug_close(cfg->debug_file);
+    cfg->debug_file = debug_open(filename);
+  } else
+    return PAM_AUTHINFO_UNAVAIL;
+
+  return PAM_SUCCESS;
+}
+
+static int slurp(int fd, size_t to_read, char **dst) {
+  char *buffer, *w;
+
+  buffer = malloc(to_read + 1);
+  if (!buffer)
+    return PAM_BUF_ERR;
+
+  w = buffer;
+  while (to_read) {
+    ssize_t r;
+
+    r = read(fd, w, to_read);
+    if (r < 0) {
+      free(buffer);
+      return PAM_SYSTEM_ERR;
+    }
+
+    if (r == 0)
+      break;
+
+    w += r;
+    to_read -= r;
+  }
+
+  *w = '\0';
+  *dst = buffer;
+  return PAM_SUCCESS;
+}
+
+static const char *ltrim(const char *s) {
+  while (isspace((unsigned char) *s))
+    s++;
+  return s;
+}
+
+static int cfg_load_defaults(cfg_t *cfg, const char *config_path) {
+  int fd, e;
+  struct stat st;
+  char *buffer = NULL, *saveptr = NULL;
+  const char *arg;
+
+  // If a config file other than the default is provided,
+  // the path must be absolute.
+  if (config_path && *config_path != '/')
+    return PAM_AUTHINFO_UNAVAIL;
+
+  fd = open(config_path ? config_path : CFG_DEFAULT_PATH,
+            O_RDONLY | O_CLOEXEC | O_NOCTTY | O_NOFOLLOW, 0);
+  if (fd == -1) {
+
+    // Only the default config file is allowed to be missing.
+    if (errno == ENOENT && !config_path)
+      return PAM_SUCCESS;
+
+    return PAM_AUTHINFO_UNAVAIL;
+  }
+
+  if (fstat(fd, &st)) {
+    e = PAM_SYSTEM_ERR;
+    goto exit;
+  }
+
+  if (!S_ISREG(st.st_mode)) {
+    e = PAM_AUTHINFO_UNAVAIL;
+    goto exit;
+  }
+
+  if (st.st_size == 0) {
+    e = PAM_SUCCESS;
+    goto exit;
+  }
+
+  if (st.st_size > CFG_MAX_FILE_SIZE) {
+    e = PAM_AUTHINFO_UNAVAIL;
+    goto exit;
+  }
+
+  e = slurp(fd, st.st_size, &buffer);
+  if (e)
+    goto exit;
+
+  arg = strtok_r(buffer, "\n", &saveptr);
+  while (arg) {
+    arg = ltrim(arg);
+
+    if (arg[0] != '\0' && arg[0] != '#') {
+      e = cfg_load_arg(cfg, arg);
```

Comment:
I think we should just be consistent with what we do for command line arguments and ignore invalid arguments. PAM is generally well-defined to do just that, [see for example RedHat](https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/6/html/managing_smart_cards/pam_configuration_files#pam-mod-args):
> Invalid arguments are generally ignored and do not otherwise affect the success or failure of the PAM module. [...]

Reply:
On this one it feels a bit weird for a security module to just disregard settings: it might weaken the system.

Regardless, consistency is good, so OK.

Also
> Most modules report errors to the /var/log/secure file.

Probably a nice TODO for later?